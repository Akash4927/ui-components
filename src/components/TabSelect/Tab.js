import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

const Styled = component => styled(component)``;

/**
 * A tab component for use within a TabSelect component.
 *  * See also:
 * [TabSelect](/components/tabselect)
 */

class Tab extends React.PureComponent {
  render() {
    const { className, children } = this.props;
    return <div className={className}>{children}</div>;
  }
}

Tab.propTypes = {
  /**
   * The label of the tab. This will appear in the TabSelect navigation section
   */
  label: PropTypes.string.isRequired,
  /**
   * A unique identifier for this tab
   */
  name: PropTypes.string.isRequired,
};

export default Styled(Tab);
