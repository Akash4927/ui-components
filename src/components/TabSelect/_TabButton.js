import React from 'react';
import styled from 'styled-components';

const TabName = styled.span``;

const getBgColor = (selected, secondary, theme) => {
  if (secondary) {
    return theme.colors.sand;
  }

  return selected ? theme.colors.white : theme.colors.lightgray;
};

const Styled = component => styled(component)`
  cursor: pointer;
  margin-bottom: -1px;
  margin-right: 5px;
  padding: 10px 20px;
  font-size: ${props => props.theme.fontSizes.large};
  outline: 0;

  ${({ selected, secondary, theme }) => `
    background-color: ${getBgColor(selected, secondary, theme)};
    border-top-left-radius: ${theme.borderRadius.soft};
    border-top-right-radius: ${theme.borderRadius.soft};
    border: 1px solid ${theme.colors.athens};
  `};

  ${props => props.selected && 'border-bottom: 1px solid transparent;'};

  ${TabName} {
    color: ${props => props.theme.colors.primary.charcoal};
    opacity: ${props => (props.selected ? 1 : 0.65)};
  }
`;

class TabButton extends React.Component {
  handeClick = ev => {
    this.props.onClick(ev, this.props.name);
  };

  render() {
    const { className, children } = this.props;
    return (
      <button onClick={this.handeClick} className={className}>
        <TabName>{children}</TabName>
      </button>
    );
  }
}

export default Styled(TabButton);
