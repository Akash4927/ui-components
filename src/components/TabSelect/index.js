export TabSelect from './TabSelect';
export Tab from './Tab';

export default from './TabSelect';
