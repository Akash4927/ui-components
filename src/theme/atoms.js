import { darken } from 'polished';

import { colors } from './colors';

// component-specific
export const atoms = {
  Alert: {
    default: {
      color: colors.black,
      background: colors.gallery,
      borderColor: darken(0.15, colors.gallery),
    },
    info: {
      color: colors.white,
      background: colors.status.info,
      borderColor: darken(0.15, colors.status.info),
    },
    success: {
      color: colors.white,
      background: colors.status.success,
      borderColor: darken(0.15, colors.status.success),
    },
    warning: {
      color: colors.white,
      background: colors.status.warning,
      borderColor: darken(0.15, colors.status.warning),
    },
    error: {
      color: colors.white,
      background: colors.status.error,
      borderColor: darken(0.15, colors.status.error),
    },
  },
  Button: {
    default: {
      color: colors.black,
      background: colors.white,
      hoverColor: colors.purple800,
      hoverBackground: darken(0.05, colors.white),
    },
    primary: {
      color: colors.white,
      background: colors.turquoise,
      hoverColor: colors.white,
      hoverBackground: darken(0.05, colors.turquoise),
    },
    danger: {
      color: colors.white,
      background: colors.status.error,
      hoverColor: colors.white,
      hoverBackground: darken(0.05, colors.status.error),
    },
    disabled: {
      color: colors.doveGray,
      background: colors.gallery,
      hoverColor: colors.doveGray,
      hoverBackground: colors.gallery,
    },
  },
};
