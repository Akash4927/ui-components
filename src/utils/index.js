import intersection from 'lodash/intersection';
import first from 'lodash/first';
import keys from 'lodash/keys';

// Return the value of a set of props against a lookup object.
// Helps when using "ternary" props, ie: <Text small bold />
// Third argument is the default value to return in the event of an `undefined`;
export const lookupValue = (props, lookup, _default) => {
  const matches = intersection(keys(props), keys(lookup));

  if (matches.length > 1) {
    throw new Error(`You have specified conflicting props: ${JSON.stringify(matches)}`);
  }

  return lookup[first(matches)] || _default;
};
