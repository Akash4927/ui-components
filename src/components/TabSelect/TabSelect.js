import React from 'react';
import styled from 'styled-components';
import { pick, map, find, get } from 'lodash';
import PropTypes from 'prop-types';

import Tab from './Tab';
import TabButton from './_TabButton';

const TabButtons = styled.div``;

const borders = props => `
    border-top-right-radius: ${props.theme.borderRadius.soft};
    border-bottom-right-radius: ${props.theme.borderRadius.soft};
    border-bottom-left-radius: ${props.theme.borderRadius.soft};
    border: 1px solid ${props.theme.colors.primary.purple100};
`;

const bordersOnlyTop = props => `
  border-top: 1px solid ${props.theme.colors.primary.purple100}
`;

const TabContent = styled.div`
  padding: 20px;
  background-color: ${props =>
    props.secondary ? 'transparent' : props.theme.colors.white};

  ${props => (props.secondary ? bordersOnlyTop(props) : borders(props))};
`;

const Styled = component => styled(component)``;

class TabSelect extends React.PureComponent {
  state = {
    selectedTab:
      // Use the first tab as the default if no selectedTab prop is specified.
      this.props.selectedTab || get(this.props.children, '[0].props.name'),
  };

  componentWillReceiveProps(nextProps) {
    // Always override user-selected tab with the default one from
    // the prop input (so that URL param can take precedence).
    if (nextProps.selectedTab !== this.props.selectedTab) {
      this.setState({ selectedTab: nextProps.selectedTab });
    }
  }

  isSelectedTab = ({ name }) => name === this.state.selectedTab;

  handleTabClick = (ev, tabName) => {
    this.setState({ selectedTab: tabName });

    if (this.props.onChange) {
      this.props.onChange(ev, tabName);
    }
  };

  render() {
    const { className, children, secondary } = this.props;
    const tabs = React.Children.map(children, child =>
      pick(child.props, ['label', 'name'])
    );

    const selected = find(children, c => this.isSelectedTab(c.props));

    return (
      <div className={className}>
        <TabButtons>
          {map(tabs, tab => (
            <TabButton
              key={tab.name}
              name={tab.name}
              selected={this.isSelectedTab(tab)}
              secondary={secondary}
              onClick={this.handleTabClick}
            >
              {tab.label}
            </TabButton>
          ))}
        </TabButtons>
        <TabContent secondary={secondary}>{selected}</TabContent>
      </div>
    );
  }
}

TabSelect.propTypes = {
  /**
   * Secondary styling for TabSelect without border around content and
   * transparent background
   */
  secondary: PropTypes.bool,
  /**
   * The tab to show on first render.
   * Supplying a new value to this prop will override the currently selected item
   */
  selectedTab: PropTypes.string,
  /**
   * Children of `TabSelect` must be a `Tab` component
   */
  children: (props, propName) => {
    let error = null;
    React.Children.forEach(props[propName], child => {
      if (child.type !== Tab) {
        error = new Error(
          `Wrong component supplied to TabSelect. Expected a <Tab />, got a ${
            child.type
          }`
        );
      }
    });
    return error;
  },
};

TabSelect.defaultProps = {
  secondary: false,
};

export default Styled(TabSelect);
