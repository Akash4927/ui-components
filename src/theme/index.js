import { colors, themeColorsAsScss } from './colors';
import { fontSizes, fontFamilies, themeFontsAsScss } from './fonts';
import { borderRadius, themeBorderRadiiAsScss } from './borders';
import { layers, themeLayersAsScss } from './layers';
import { spacing, themeSpacingsAsScss } from './spacings';
import { overlayIconSize, textColor, themeMiscVarsAsScss } from './misc';
import { boxShadow } from './box-shadows';
import { atoms } from './atoms';

export function themeVarsAsScss() {
  const themeVariables = []
    .concat(themeColorsAsScss())
    .concat(themeLayersAsScss())
    .concat(themeFontsAsScss())
    .concat(themeBorderRadiiAsScss())
    .concat(themeSpacingsAsScss())
    .concat(themeMiscVarsAsScss());
  return `${themeVariables.join('; ')};`;
}

const theme = {
  // Component-specific
  atoms,

  // Box shadows
  boxShadow,

  // Borders
  borderRadius,

  // Colors
  colors,

  // Fonts
  fontFamilies,
  fontSizes,

  // z-index layers
  layers,

  // Spacings
  spacing,

  // Misc
  overlayIconSize,
  textColor,
};

export default theme;
