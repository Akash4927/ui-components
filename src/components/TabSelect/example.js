import React from 'react';

import { Example, Info } from '../../utils/example';
import TabSelect from './TabSelect';
import Tab from './Tab';

export default function TabSelectExample() {
  return (
    <div>
      <Example>
        <Info>Default</Info>
        <TabSelect>
          <Tab label="A" name="a">
            Tab A
          </Tab>
          <Tab label="B" name="b">
            Tab B
          </Tab>
          <Tab label="C" name="c">
            Tab C
          </Tab>
        </TabSelect>
      </Example>
      <Info>Secondary</Info>
      <Example>
        <TabSelect secondary>
          <Tab label="A" name="a">
            Tab A
          </Tab>
          <Tab label="B" name="b">
            Tab B
          </Tab>
          <Tab label="C" name="c">
            Tab C
          </Tab>
        </TabSelect>
      </Example>
    </div>
  );
}
