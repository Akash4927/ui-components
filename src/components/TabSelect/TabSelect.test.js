import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import 'jest-styled-components';
import { withTheme } from '../../utils/theme';

import TabSelect from './TabSelect';
import Tab from './Tab';
import TabButton from './_TabButton';

describe('<Tab Select />', () => {
  describe('snapshots', () => {
    it('renders', () => {
      const tree = renderer
        .create(
          withTheme(
            <TabSelect>
              <Tab label="A" name="a">
                Tab A
              </Tab>
              <Tab label="B" name="b">
                Tab B
              </Tab>
              <Tab label="C" name="c">
                Tab C
              </Tab>
            </TabSelect>
          )
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('renders with a selectedTab', () => {
      const tree = renderer
        .create(
          withTheme(
            <TabSelect selectedTab="b">
              <Tab label="A" name="a">
                Tab A
              </Tab>
              <Tab label="B" name="b">
                Tab B
              </Tab>
              <Tab label="C" name="c">
                Tab C
              </Tab>
            </TabSelect>
          )
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
  it('changes tabs on click', () => {
    const tabs = mount(
      withTheme(
        <TabSelect>
          <Tab label="A" name="a">
            Tab A
          </Tab>
          <Tab label="B" name="b">
            Tab B
          </Tab>
          <Tab label="C" name="c">
            Tab C
          </Tab>
        </TabSelect>
      )
    );

    tabs
      .find(TabButton)
      .at(1)
      .simulate('click');
    expect(tabs.find(Tab).text()).toEqual('Tab B');
  });
});
