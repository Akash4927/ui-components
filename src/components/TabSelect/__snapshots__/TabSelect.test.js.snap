// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tab Select /> snapshots renders 1`] = `
.c0 {
  cursor: pointer;
  margin-bottom: -1px;
  margin-right: 5px;
  padding: 10px 20px;
  font-size: 22px;
  outline: 0;
  background-color: #fff;
  border-top-left-radius: 4px;
  border-top-right-radius: 4px;
  border: 1px solid #e2e2ec;
  border-bottom: 1px solid transparent;
}

.c0 .c1 {
  color: #32324b;
  opacity: 1;
}

.c2 {
  cursor: pointer;
  margin-bottom: -1px;
  margin-right: 5px;
  padding: 10px 20px;
  font-size: 22px;
  outline: 0;
  background-color: #f4f4f4;
  border-top-left-radius: 4px;
  border-top-right-radius: 4px;
  border: 1px solid #e2e2ec;
}

.c2 .c1 {
  color: #32324b;
  opacity: 0.65;
}

.c3 {
  padding: 20px;
  background-color: #fff;
  border-top-right-radius: 4px;
  border-bottom-right-radius: 4px;
  border-bottom-left-radius: 4px;
  border: 1px solid #e2e2ec;
}

<div
  className=""
>
  <div
    className=""
  >
    <button
      className="c0"
      onClick={[Function]}
    >
      <span
        className="c1 "
      >
        A
      </span>
    </button>
    <button
      className="c2"
      onClick={[Function]}
    >
      <span
        className="c1 "
      >
        B
      </span>
    </button>
    <button
      className="c2"
      onClick={[Function]}
    >
      <span
        className="c1 "
      >
        C
      </span>
    </button>
  </div>
  <div
    className="c3"
  >
    <div
      className=""
    >
      Tab A
    </div>
  </div>
</div>
`;

exports[`<Tab Select /> snapshots renders with a selectedTab 1`] = `
.c2 {
  cursor: pointer;
  margin-bottom: -1px;
  margin-right: 5px;
  padding: 10px 20px;
  font-size: 22px;
  outline: 0;
  background-color: #fff;
  border-top-left-radius: 4px;
  border-top-right-radius: 4px;
  border: 1px solid #e2e2ec;
  border-bottom: 1px solid transparent;
}

.c2 .c1 {
  color: #32324b;
  opacity: 1;
}

.c0 {
  cursor: pointer;
  margin-bottom: -1px;
  margin-right: 5px;
  padding: 10px 20px;
  font-size: 22px;
  outline: 0;
  background-color: #f4f4f4;
  border-top-left-radius: 4px;
  border-top-right-radius: 4px;
  border: 1px solid #e2e2ec;
}

.c0 .c1 {
  color: #32324b;
  opacity: 0.65;
}

.c3 {
  padding: 20px;
  background-color: #fff;
  border-top-right-radius: 4px;
  border-bottom-right-radius: 4px;
  border-bottom-left-radius: 4px;
  border: 1px solid #e2e2ec;
}

<div
  className=""
>
  <div
    className=""
  >
    <button
      className="c0"
      onClick={[Function]}
    >
      <span
        className="c1 "
      >
        A
      </span>
    </button>
    <button
      className="c2"
      onClick={[Function]}
    >
      <span
        className="c1 "
      >
        B
      </span>
    </button>
    <button
      className="c0"
      onClick={[Function]}
    >
      <span
        className="c1 "
      >
        C
      </span>
    </button>
  </div>
  <div
    className="c3"
  >
    <div
      className=""
    >
      Tab B
    </div>
  </div>
</div>
`;
